### PR TITLE
Support overlapping tiles

### LINKED ISSUE #22: "Uniform tiling" in scope
I have an image and I want to decompose it into tiles. Such that
* All tiles have the same fixed size
* Tiles should overlap as little as possible and as uniformly as possible. E.g.
```julia
img = 1:9
size = 4
tiles(img, size) = [1:4, 5:8, 6:9] # bad last tile has huge overlap
tiles(img, size) = [1:4, 4:7, 6:9] # good
tiles(img, size) = [1:4, 3:6, 6:9] # also good
```
Is that in scope of this package? Suggestions for another package, where this could live?

### PR DESCRIPTION
**Pull request by jw3126**
Close #22 
This needs at least docs and more tests. But before I go into that, @johnnychen94 what do you think about the API/design of this PR?

### CODE REVIEW

**Comment by johnnychen94** [comment]
As a prototype, there's no doubt that this version works for your purpose. I have thought about the enhancement of this package but haven't dig into all the possibilities and implementation details. Here are some of my thoughts.

---

Codes in the current version seems a bit coupled together and might be very hard to extend. If it is possible, I feel it would be better and cleaner to implement a new Range type for the Balanced version, and reuse StepRange for the Fixed version. And keep the previous struct type as possible as we can.

```julia
struct TileIterator{N,C} <: AbstractArray{NTuple{N, UnitRange{Int}}, N}
    covers1d::C
end
```
IIUC `covers1d` here is a tuple of vectors, which isn't great in performance. We should make this vector lazily generated. I think that's why the previous struct type is designed that way.

---

Generally, I don't think it's a good direction to make it an array type. For simple cases like this, we could have efficient getindex implementation, but that's not always possible for other cases. Make it an iterator would be more efficient and extensible, but yes, it requires some effort to implement it.

---

I'm not sure if we should add `stride` keyword here, I feel it'd be better to

```julia
TileIterator(axes::Indices, ::IterationStrategy)
```

size and stride are just meta info to iteration strategy. I prefer to let `TileIterator` to normalize the axes input/output and let `IterationStrategy` to handle the dirty and complex iteration implementation.

Does every strategy has a stride meta? I don't know, I don't feel it good to assume that. We assume iteration along the column direction here, it might be good to assume that we also want to support iteration along row direction in the future. How would we solve that, adding a new `columnmajor=true` keyword? Probably, but that would exponentially increase the complexity.

In general, I want to get to a design where Iteration Strategy are pluggable and composable.

**Comment by jw3126** [comment]
Thanks a lot for your comments. I agree with most of what you said, but have a different view on some points:

> Codes in the current PR version seems a bit coupled together and might be very hard to extend. If it is possible, I feel it would be better and cleaner to implement a new Range type for the Balanced version, and reuse StepRange for the Fixed version. And keep the previous struct type as possible as we can.
> 
> ```julia
> struct TileIterator{N,C} <: AbstractArray{NTuple{N, UnitRange{Int}}, N}
>     covers1d::C
> end
> ```
> 
> IIUC `covers1d` here is a tuple of vectors, which isn't great in performance. We should make this vector lazily generated. I think that's why the previous struct type is designed that way.

There seems to be a misunderstanding `covers1d` is a tuple of `AbstactVector{UnitRange{Int}}`. It can be a tuple of `Vector`, but that is not the default. The default is lazy. Either `StepRange` or `QuantizedRange` in the balanced case. I think this is very flexible and easy to extend. I did not look into performance yet.
I could imagine, that a second field is needed for faster iteration, but I would like to avoid that if possible.
> struct TileIterator{...}
>     covers1d::C
>     _fast_iteration_implementation_detail::F
> end

> 
> Generally, I don't think it's a good direction to make it an array type. For simple cases like this, we could have efficient getindex implementation, but that's not always possible for other cases. Make it an iterator would be more efficient and extensible, but yes, it requires some effort to implement it.

I think in the above situation it is very natural to make an `AbstractArray` and `getindex` can be fast. And I think it will be easy for any tiling that is a product of 1d cases. If we have a case that is not a product, it should use another type. Something like
```julia
struct ExoticTileIterator <: NotAnAbstractArray
...
end
```
Maybe we should use `tileiterator(axes, strategy)` and it will create a `ProductTiling <: AbstractArray` or `MyExoticTiling <: NotAnArray`.

> 
> I'm not sure if we should add `stride` keyword here, I feel it'd be better to
> 
> ```julia
> TileIterator(axes::Indices, ::IterationStrategy)
> ```
> 
> size and stride are just meta info to iteration strategy. I prefer to let `TileIterator` to normalize the axes input/output and let `IterationStrategy` to handle the dirty and complex iteration implementation.

Excellent point I agree.

> 
> Does every strategy has a stride meta? I don't know, I don't feel it good to assume that. We assume iteration along the column direction here, it might be good to assume that we also want to support iteration along row direction in the future. How would we solve that, adding a new `columnmajor=true` keyword? Probably, but that would exponentially increase the complexity.
> 

I fully agree here, is better to go with your `::IterationStrategy` suggestion.

> In general, I want to get to a design where iteration strategies are pluggable and composable.

Again I fully agree. I am however not sure, what pluggable and composable mean here concretely.

**Comment by johnnychen94** [comment]
I'm a little bit with my own work here and I'm sorry that I can't give any detailed feedback on the code at this moment. It looks like you clearly know what should be done. My major concern about this PR and the changes is the performance overhead and extensibility. If that's done and I'm good with that.

---

> > IIUC covers1d here is a tuple of vectors, which isn't great in performance. We should make this vector lazily generated. I think that's why the previous struct type is designed that way.
>
> There seems to be a misunderstanding covers1d is a tuple of AbstactVector{UnitRange{Int}}. It can be a tuple of Vector, but that is not the default. The default is lazy. Either StepRange or QuantizedRange in the balanced case. I think this is very flexible and easy to extend. I did not look into performance yet.

Hmmm, I don't quite get it here because the constructor tells me this.

```julia
function TileIterator(covers1d::NTuple{N, AbstractVector{UnitRange{Int}}}) where {N}
    C = typeof(covers1d)
    return TileIterator{N, C}(covers1d)
end
```

This representation might bring some overhead and I'd like to avoid that overhead. I probably am wrong here but still, a performance benchmark is needed to support this change.

> Maybe we should use tileiterator(axes, strategy) and it will create a ProductTiling <: AbstractArray or MyExoticTiling <: NotAnArray.

Yeah, perhaps. It seems like this doesn't need to be included in this PR.


> > In general, I want to get to a design where iteration strategies are pluggable and composable.
> Again I fully agree. I am however not sure, what pluggable and composable mean here concretely.

By composable I mean two iteration strategy can be composed together into another strategy. For example, `Fixed() ∘ RowMajor()` or something like this. Again, this doesn't need to be included in this PR but it would be better that we consider this possibility and extensibility.

**Comment by jw3126** [comment]
> Hmmm, I don't quite get it here because the constructor tells me this.
> 
> ```julia
> function TileIterator(covers1d::NTuple{N, AbstractVector{UnitRange{Int}}}) where {N}
>     C = typeof(covers1d)
>     return TileIterator{N, C}(covers1d)
> end
> ```
> 
> This representation might bring some overhead and I'd like to avoid that overhead. I probably am wrong here but still, a performance benchmark is needed to support this change.

This constructor is a bit low level. If you ask by whatever API for the `balanced` case, you would get
```julia
typeof(covers1d) = Tuple{CoveredRange{QuantizedRange....
```
which is a bitstype and completly lazy. Did you worry because you thought e.g. this would allocate a `Vector` or are you worrying that the compiler does not properly inline away all the abstraction?

**Comment by johnnychen94** [comment]
Ha, I get your point here; didn't realize that `CoveredRange <: AbstractVector`. I was worried about the additional overhead during `getindex(::Vector, i)` but since this is lazily generated, it looks good.

I was worried about the overhead during iteration, for example:

```julia
julia> titr = TileIterator((1:7, 3:6), tilesize=(2, 2), stride=Balanced((3, 2)));

julia> @btime getindex(titr, 2, 2)
  13.920 ns (0 allocations: 0 bytes)
(3:4, 5:6)

julia> titr = TileIterator((1:7, 3:6), tilesize=(2, 2), stride=Fixed((3, 2)));

julia> @btime getindex(titr, 2, 2)
  0.045 ns (0 allocations: 0 bytes)
(4:5, 5:6)
```

great that this has zero allocation. So yeah, this implementation is quite efficient already. Perhaps there's still some room to get `getindex` for Balanced faster.

---

This looks good to me now, a really nice play with the types and dispatches. I guess we just need to leave some room for future extensibility as we discussed. Then add docs and tests and I feel it could be good to merge.

File: src/tileiterator.jl
Comment (jw3126) [inline src/tileiterator.jl:72]: I thought, lets be conservative for now. I just added these two strategies. They are enough to cover the funcionality on `master` as well as my balanced tile size use case.

Comment (jw3126) [reply src/tileiterator.jl:72]: A different PR could implement a strategy that sets the stride to `(1,1,1...)` giving the behavior sketched here:
https://github.com/JuliaImages/ImageFiltering.jl/pull/155#issuecomment-706649346

Comment (jw3126) [reply src/tileiterator.jl:72]: For more advanced strides, we might want to do
`tileiterator(axes, UnitStride())[begin:2:end, :, begin+2:4:end-2]`

File: src/tileiterator.jl
Comment (johnnychen94) [inline src/tileiterator.jl:59]: The name `tileiterator` gives me an indication that it might not be type stable. Could we just make `tileiterator` the outer constructor of `TileIterator` here?

Comment (jw3126) [reply src/tileiterator.jl:59]: You mean I should delete the `tileiterator` function and instead have `TileIterator(axes, strategy)`?

Comment (johnnychen94) [reply src/tileiterator.jl:59]: Yes, it looks like there's no need for `tileiterator` here.

File: src/tileiterator.jl
Comment (johnnychen94) [reply src/tileiterator.jl:95]: This works but might not be the most efficient way because indexing with `[begin:2:end, :, begin+2:4:end-2]` allocates a new array. No need to implement in this PR, though.

File: src/tileiterator.jl
Comment (jw3126) [reply src/tileiterator.jl:59]: Ok done. Can you attach the `hacktoberfest-accepted` label to this PR? See [here](https://hacktoberfest.digitalocean.com/hacktoberfest-update) for background.

**Comment by johnnychen94** [comment]
Probably we could just switch to use Travis for the Windows platform and sunset appveyor.

File: src/tileiterator.jl
Comment (johnnychen94) [inline src/tileiterator.jl:71]: Would you mind also adding the docs for `TileIterator(axes, tilesize)`? It would be good to also have some words on what `strategy` is.

File: test/runtests.jl
Comment (johnnychen94) [inline test/runtests.jl:32]: Should also add some test cases for 2d cases; those are currently live in doctest but it'd better to also have them here.

**Comment by jw3126** [comment]
I did remove appveyor.yml and travis passes on windows. Probably need to deactivate appveyor on the appveyor website or something?

File: test/runtests.jl
Comment (johnnychen94) [inline test/runtests.jl:32]: As a special case, it would be very helpful to add some tests for other range types, e.g., `OffsetArrays.IdOffsetRange` and `Base.IdentityUnitRange`; those are usually very good test cases to catch exceptions.

```julia
# unsure if this works
using OffsetArrays
A = zeros(10)
AO = OffsetArray(A, OffsetArrays.Origin(0))

titr = TileIterator(axes(AO), RelaxStride((3, )))
@test titr == [(0:2,), (3:5,), (6:8,), (9:9,)]
```

File: src/tileiterator.jl
Comment (johnnychen94) [inline src/tileiterator.jl:81]: Curious to ask, is there a way to specify `RelaxLastTile` for the first dimension, while specifying `RelaxStride` for the second dimension? I guess this can be achieved somehow with `map`?

If this is actually supported, might be helpful to add it as an example, too.

**Review by johnnychen94 — COMMENTED** [review]
The current version looks good to me; probably the last round review. More docs and tests might be helpful, though.

**Comment by johnnychen94** [comment]
> I did remove appveyor.yml and travis passes on windows. Probably need to deactivate appveyor on the appveyor website or something?

Will remove that once this PR is merged.

File: src/tileiterator.jl
Comment (jw3126) [reply src/tileiterator.jl:81]: There is no API for that. Would not hard to support this though.

File: src/tileiterator.jl
Comment (jw3126) [reply src/tileiterator.jl:132]: Yes you are right this allocates currently, we can fix it, when it becomes a practical obstacle.

File: src/tileiterator.jl
Comment (jw3126) [reply src/tileiterator.jl:71]: done

File: test/runtests.jl
Comment (jw3126) [reply test/runtests.jl]: done

Comment (jw3126) [reply test/runtests.jl]: done

**Review by johnnychen94 — APPROVED** [review]
LGTM, I think it's good to merge once the test passes. This is not a breaking change, but I'll bump to v0.3.0 since it drops Julia v0.7 support.

ping @timholy in case he has some thoughts about the direction of this PR.

**Comment by jw3126** [comment]
Thanks @johnnychen94 for the detailed review and feedback.

**Comment by timholy** [comment]
I've got this on my calendar to review tomorrow, sorry for the delay.

File: src/tileiterator.jl
Comment (timholy) [inline src/tileiterator.jl:108]:
```suggestion
```
(redundant with export in TiledIteration.jl)

File: src/tileiterator.jl
Comment (timholy) [inline src/tileiterator.jl:136]:
```suggestion
```

File: src/tileiterator.jl
Comment (timholy) [inline src/tileiterator.jl:141]:
```suggestion
Tiling strategy, that permits the size of the last tiles along each dimension to be smaller
```

File: src/tileiterator.jl
Comment (timholy) [inline src/tileiterator.jl:142]:
```suggestion
than `tilesize` if needed. All other tiles are of size `tilesize`.
```

File: src/tileiterator.jl
Comment (timholy) [inline src/tileiterator.jl:186]: Is the return type assertion doing anything? If so then you have an inference problem here, but I'm not seeing it by eye.

File: test/runtests.jl
Comment (timholy) [inline test/runtests.jl:50]: Maybe
```julia
IdentityUnitRange = OffsetArrays.IdentityUnitRange
```
which uses `Base.Slice` on Julia 1.0.

**Review by timholy — COMMENTED** [review]
This is really great! In the future we may also want to move some of the tile-padding logic from ImageFiltering here, and this will be a good foundation.

File: src/tileiterator.jl
Comment (jw3126) [reply src/tileiterator.jl:24]: How can this be called?

File: src/tileiterator.jl
Comment (jw3126) [reply src/tileiterator.jl:37]: Oh nice, I was not aware of `RoundingIntegers`. I think I prefer `RoundedRange` for now. If I have time, I will add some polish to `RoundingIntegers` e.g. add a `Project.toml`. And use it instead.

File: src/tileiterator.jl
Comment (jw3126) [reply src/tileiterator.jl:184]: No it is just "documentation" should I drop it?

File: test/runtests.jl
Comment (jw3126) [reply test/runtests.jl:50]: done

File: src/tileiterator.jl
Comment (jw3126) [reply src/tileiterator.jl:24]: So I think this cannot get called when using only public API.

Comment (jw3126) [reply src/tileiterator.jl:24]: I disallowed nothing. It was there only for "historic" reasons.

File: src/tileiterator.jl
Comment (timholy) [reply src/tileiterator.jl:184]: Maybe put a `#` in front of it, just to be clear it's not functional.

File: test/runtests.jl
Comment (timholy) [reply test/runtests.jl:50]: Needs a push?

**Comment by timholy** [comment]
There may be something weird going on. E.g., https://github.com/JuliaArrays/TiledIteration.jl/pull/23/commits/75b2a90ecf394cdab1a452c60929c770c7bbd0cc doesn't match the commit-summary.

**Comment by jw3126** [comment]
Yeah I messed something up. I think we can just squash this PR or should I try to clean the history?

File: src/tileiterator.jl
Comment (jw3126) [reply src/tileiterator.jl]: I just removed it.

**Comment by timholy** [comment]
Thanks again, terrific to have this!

**Comment by timholy** [comment]
There's nothing breaking about this, right?

Seems a little weird to release such a major improvement as 0.2.6, but OTOH not having to do version bumps in dependencies has advantages.

**Comment by timholy** [comment]
Packages that would need [compat] bumps are just ImageFiltering, ImageMorphology, and Images (of those in General, anyway). Not bad at all. If we do want to increase the version number, perhaps we should contemplate 1.0?

**Comment by jw3126** [comment]
Yeah, I think this is not breaking. I would go with `0.2.6`. For 1.0 I think we might want to gather some more experience with the `TileIterator(axes, strategy)` API.

**Comment by johnnychen94** [comment]
The only reason to bump to v0.3.0 is that Julia 0.7 support is dropped here.

**Comment by timholy** [comment]
Oh right, we'll have to bump in fact. OK, let's release this as 0.3.

**Pull request merged**
